### PR TITLE
feat(react-motions): add grouped motions

### PR DESCRIPTION
**Pull request by layershifter**
## New Behavior

This PR adds support for grouped motions to both `createMotionComponent()` & `createPresenceComponent()`:

```ts
const FadeFastGrowSlow = createMotionComponent([
  {
    keyframes: [{ opacity: 0 }, { opacity: 1 }],
    duration: 200,
    easing: 'easeIn',
  },
  {
    keyframes: [{ transform: 'scale(0)' }, { transform: 'scale(1)' }],
    duration: 500 /* 💡 note the different duration */,
    easing: 'cubic-bezier(0.42, 0, 0.58, 1)' /* 💡 note the different easing */,
  },
]);
```

This mimics CSS following CSS:

```css
@keyframes fade {
  from { opacity: 0 }
  to { opacity: 1 }
}

@keyframes scale {
  from { transform: scale(0) }
  to { transform: scale(1) }
}

.motion {
  animation-name: fade, scale;
  animation-duration: 200ms, 500ms;
  animation-timing-function: easeIn, cubic-bezier(0.42, 0, 0.58, 1);
}
```

## Notes

- Animation logic have been refactored, now `createMotionComponent()` & `createPresenceComponent()` share it
- New docs have been added

## Related Issue(s)

Partially implements #30547.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

✅ No changes found
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">33</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">30</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/Avatar-mount_base.html">656</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/Avatar-mount.html">635</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/Button-mount_base.html">297</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/Button-mount.html">300</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/Field-mount_base.html">1113</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/Field-mount.html">1093</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProvider-mount_base.html">697</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProvider-mount.html">700</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-mount_base.html">76</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-mount.html">80</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">33</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">30</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">78</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">79</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/MakeStyles-mount_base.html">871</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/MakeStyles-mount.html">866</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/Persona-mount_base.html">1739</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/Persona-mount.html">1716</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/SpinButton-mount_base.html">1397</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/SpinButton-mount.html">1372</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/SwatchPicker-mount_base.html">1511</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31487/perf-test-react-components/SwatchPicker-mount.html">1541</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: packages/react-components/react-motions-preview/stories/CreatePresenceComponent/MotionArrays.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
+import { makeStyles, tokens, Label, Slider, useId, Checkbox } from '@fluentui/react-components';
+import { createPresenceComponent, motionTokens } from '@fluentui/react-motions-preview';
+import type { MotionImperativeRef } from '@fluentui/react-motions-preview';
+import * as React from 'react';
+
+import description from './PresenceAppear.stories.md';
```
Comment (layershifter) [inline packages/react-components/react-motions-preview/stories/CreatePresenceComponent/MotionArrays.stories.tsx:6]: FIX: proper description

File: packages/react-components/react-motions-preview/src/factories/createPresenceComponent.ts [hunk 2]
```diff
@@ -81,37 +81,37 @@ export function createPresenceComponent(motion: PresenceMotion | PresenceMotionF
...
-          animation.finish();
+          handle.finish();
           return;
         }
 
-        animationRef.current = animation;
-        animation.onfinish = visible ? onEnterFinish : onExitFinish;
+        handle.finished.then(() => {
+          // TODO: FIXME
+          const event = new AnimationPlaybackEvent('');
```
Comment (layershifter) [inline packages/react-components/react-motions-preview/src/factories/createPresenceComponent.ts:99]: FIX: Break in a separate PR

File: change/@fluentui-react-motions-preview-2a7ccd1a-4e03-4c37-812f-a8abe7a8fde4.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-motions-preview-2a7ccd1a-4e03-4c37-812f-a8abe7a8fde4.json:1]: ## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31487&commitId=774648dec107ea95bcdf5e592370c4315cd56237&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask - RTL.normal.chromium.png|3| Changed|
|Avatar Converged.badgeMask.normal.chromium.png|1| Changed|
|Avatar Converged.size+image+badge.normal.chromium.png|0| Added|

</details>


  <details><summary><b>SwatchPicker Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|SwatchPicker Converged.Shape.default.chromium.png|0| Removed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Review by robertpenner — APPROVED** [review]

File: packages/react-components/react-motions-preview/src/types.ts [hunk 4]
```diff
@@ -1,13 +1,19 @@
...
 
 export type PresenceMotion = {
-  enter: AtomMotion;
-  exit: AtomMotion;
+  enter: AtomMotion | AtomMotion[];
+  exit: AtomMotion | AtomMotion[];
 };
 
-export type AtomMotionFn = (element: HTMLElement) => AtomMotion;
+export type AtomMotionFn = (element: HTMLElement) => AtomMotion | AtomMotion[];
```
Comment (ling1726) [inline packages/react-components/react-motions-preview/src/types.ts:8]: just an idea for future: it would be nice if this function also `isReducedMotion`

File: packages/react-components/react-motions-preview/stories/CreateMotionComponent/MotionArrays.stories.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+`createMotionComponent()` supports arrays of `AtomMotion` objects. This is useful when you want to animate properties with _different_ durations, easings, etc.
+
+```ts
+const FadeFastGrowSlow = createMotionComponent([
```
Comment (ling1726) [inline packages/react-components/react-motions-preview/stories/CreateMotionComponent/MotionArrays.stories.md:4]: I think examples that also involve synchronizing delays would be useful to users

**Review by ling1726 — APPROVED** [review]

File: packages/react-components/react-motions-preview/src/utils/animateAtoms.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        animation.playbackRate = rate;
+      });
+    },
+    set onfinish(callback: () => void) {
+      Promise.all(animations.map(animation => animation.finished))
+        .then(() => {
+          callback();
+        })
+        .catch(() => {
+          // Ignores "DOMException: The user aborted a request" that appears if animations are cancelled
```
Comment (robertpenner) [inline packages/react-components/react-motions-preview/src/utils/animateAtoms.ts:46]: Could we filter out this specific exception and re-throw the others? Swallowing all the exceptions could make some rare cases harder to debug.

File: packages/react-components/react-motions-preview/src/types.ts [hunk 7]
```diff
@@ -1,13 +1,19 @@
 export type AtomMotion = { keyframes: Keyframe[] } & KeyframeEffectOptions;
 
 export type PresenceMotion = {
-  enter: AtomMotion;
-  exit: AtomMotion;
+  enter: AtomMotion | AtomMotion[];
+  exit: AtomMotion | AtomMotion[];
```
Comment (robertpenner) [inline packages/react-components/react-motions-preview/src/types.ts:5]: `AtomMotion | AtomMotion[]` occurs 9 times in this diff. While there are benefits to being explicit and literal (it's easy to read), DRYing it up in a type could help ensure future APIs support both the singular and plural.

File: packages/react-components/react-motions-preview/stories/CreateMotionComponent/MotionArrays.stories.md (hunk 5 above)
Comment (robertpenner) [reply packages/react-components/react-motions-preview/stories/CreateMotionComponent/MotionArrays.stories.md:4]: An important delay use case in the near future: we want a **collapse** motion that expands the height first, then starts the fade-in after a delay (there could be some overlap). The exit would fade out, then collapse the height. This is useful for adding/removing items from lists.

File: packages/react-components/react-motions-preview/src/types.ts (hunk 4 above)
Comment (robertpenner) [reply packages/react-components/react-motions-preview/src/types.ts:8]: I also would like to see reduced motion be parameterized throughout. It could be an optional that defaults to the browser's setting, but overridable for testing. This would enable use cases like displaying both the regular and reduced motion versions of a motion component side-by-side.

Comment (layershifter) [reply packages/react-components/react-motions-preview/src/types.ts:8]: I would say that it depends if we want to make `isReducedMotion` declarative or not. The original proposal was to have:

```ts
const atom = {
  keyframes: [],
  duration: 500,
  reducedMotion: {
    keyframes: [],
    duration: 100,
  }
}
```

Anyway this could be added later 😉 

> It could be an optional that defaults to the browser's setting, but overridable for testing. This would enable use cases like displaying both the regular and reduced motion versions of a motion component side-by-side.

We can implement this as a React context how it was done in v0 & in some other libraries. Anyway, that's something that we can address separately 🐱

File: packages/react-components/react-motions-preview/stories/CreateMotionComponent/MotionArrays.stories.md (hunk 5 above)
Comment (layershifter) [reply packages/react-components/react-motions-preview/stories/CreateMotionComponent/MotionArrays.stories.md:4]: @ling1726 Is there something particular that you would like to see as an example?

> An important delay use case in the near future: we want a collapse motion that expands the height first, then starts the fade-in after a delay (there could be some overlap). The exit would fade out, then collapse the height. This is useful for adding/removing items from lists.

@robertpenner well, it's what this PR unblocks 👍

File: packages/react-components/react-motions-preview/src/utils/animateAtoms.ts (hunk 6 above)
Comment (layershifter) [reply packages/react-components/react-motions-preview/src/utils/animateAtoms.ts:48]: Done, good call 👍

File: packages/react-components/react-motions-preview/src/types.ts (hunk 7 above)
Comment (layershifter) [reply packages/react-components/react-motions-preview/src/types.ts:5]: I will keep it as it is for now as it's more explicit. We can reconsider this later, it won't be a breaking change.

File: packages/react-components/react-motions-preview/stories/CreateMotionComponent/MotionArrays.stories.md (hunk 5 above)
Comment (robertpenner) [reply packages/react-components/react-motions-preview/stories/CreateMotionComponent/MotionArrays.stories.md:4]: @layershifter Yes, of course. The use case I mentioned happens to answer your question to Ling.
